keps/sig-api-machinery/34-storage-hash.md

```diff
@@ -0,0 +1,181 @@
+---
+kep-number: 31
+title: Enabling clients to tell if resource endpoints serve the same set of objects
+authors:
+  - "@xuchao"
+owning-sig: sig-api-machinery
+reviewers:
+  - "@deads2k"
+  - "@lavalamp"
+approvers:
+  - "@deads2k"
+  - "@lavalamp"
+creation-date: 2018-10-12
+last-updated: 2018-xx-xx
+status: provisional
+---
+
+# Enabling clients to tell if resource endpoints serve the same set of objects
+
+## Table of Contents
+
+      * [Summary](#summary)
+      * [Motivation](#motivation)
+            * [Broken forward compatibility](#broken-forward-compatibility)
+            * [Inaccurate resource quota counting](#inaccurate-resource-quota-counting)
+            * [Inefficiency of API-discovery based dynamic clients](#inefficiency-of-api-discovery-based-dynamic-clients)
+      * [Goals](#goals)
+      * [Proposal](#proposal)
+         * [API changes to the discovery API](#api-changes-to-the-discovery-api)
+         * [Implementation details](#implementation-details)
+         * [Risks and Mitigations](#risks-and-mitigations)
+      * [Graduation Criteria](#graduation-criteria)
+      * [Alternatives](#alternatives)
+
+## Summary
+
+We propose to expand the discovery API to enable clients to tell if resource
+endpoints (e.g., `extensions/v1beta1/replicaset` and `apps/v1/replicaset`) are
+referring the same set of objects.
+
+## Motivation
+
+Sometimes clients want to process every object in the cluster (examples below).
+Lack of visibility makes it hard for clients to understand which resources they
+need to process in that case.
+
+#### Broken forward compatibility
+
+For example, if we configure an admission webhook to validate all replicasets,
+one would set the webhook configuration this way:
+```yaml
+rules:
+  - apiGroups:
+    - "extensions,apps"
+    apiVersions:
+    - "*"
+    resources:
+    - replicasets
+```
+
+If in a future Kubernetes release, replicasets are accessible via
+`fancy-apps/v1` as well, requests sent to `fancy-apps/v1` will be left
+unchecked. Note that the admins of cluster upgrades are not necessarily the
+admins of the admission webhooks, so it is not always possible to coordinate
+cluster upgrades with webhook configuration upgrades.
+
+#### Inaccurate resource quota counting
+
+Although `extensions/v1beta1/replicasets` and `apps/v1/replicasets` refer to the
+same set of objects, they have separate resource quotas. User can bypass the 
+resource quota of one endpoint by using the other endpoint.
+
+#### Inefficiency of API-discovery based dynamic clients
+
+For example, the Kubernetes [garbage collector][] watches all discoverable
+resource endpoints and thus wasting bandwidth. As another example, the [storage
+migrator][] migrates the same objects multiple times if they are served via
+multiple endpoints.
+
+[garbage collector]:https://github.com/kubernetes/kubernetes/tree/master/pkg/controller/garbagecollector
```

Comment:
GC today doesn't even collapse versions of the same groupresource, does it?  If you're worried, you could get a lot of improvement just with that.

Reply:
I removed this example. GC's internal graph is keyed by UIDs, so it is collapsed already.

Network bandwidth is not a concern today either, because GC is using the shared informer, unless we want to cut GC into its own binary in the future.

Reply:
We can revisit the GC case if we decide to pursue https://github.com/kubernetes/kubernetes/issues/72192#issuecomment-448861055.